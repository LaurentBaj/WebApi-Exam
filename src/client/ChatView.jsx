import * as React from "react";
import { useState } from "react";

export function ChatView({ onSendMessage, chatLog, member }) {
  const [message, setMessage] = useState("");

  function handleSubmitChatMessage(e) {
    e.preventDefault();
    onSendMessage(message);
    setMessage("");
  }

  return (
    <>
      <footer>
        <form onSubmit={handleSubmitChatMessage}>
          <input
            type="text"
            autoFocus={true}
            value={message}
            onChange={(e) => setMessage(e.target.value)}
          />
          <button>Submit</button>
        </form>
      </footer>
      <main>
        <div className={"output"}>
          {chatLog.map((message, index) => (
            <div key={index}>
              {member.firstName} wrote: "{message}"
            </div>
          ))}
        </div>
      </main>
    </>
  );
}
