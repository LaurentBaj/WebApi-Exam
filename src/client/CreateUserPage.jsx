import React, { useState } from "react";
import { InputField } from "./components/InputField";
import { useHistory } from "react-router";

export function CreateUserPage({ userApi }) {
  const [firstName, setFistName] = useState("");
  const [lastName, setLastName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const history = useHistory();

  async function submit(e) {
    e.preventDefault();
    history.push("/");
    await userApi.createUser({ firstName, lastName, email, password });
  }

  return (
    <form onSubmit={submit}>
      <h1>Create new user</h1>
      <InputField
        label={"First Name"}
        value={firstName}
        onChangeValue={setFistName}
      />
      <InputField
        label={"Last Name"}
        value={lastName}
        onChangeValue={setLastName}
      />
      <InputField label={"Email"} value={email} onChangeValue={setEmail} />
      <InputField
        label={"Password"}
        value={password}
        onChangeValue={setPassword}
        type="password"
      />
      <button>Submit</button>
    </form>
  );
}
