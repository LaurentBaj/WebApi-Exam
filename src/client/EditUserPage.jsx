import React, { useState } from "react";
import { LoadingView } from "./components/LoadingView";
import { InputField } from "./components/InputField";
import { useLoading } from "./useLoading";
import { ErrorView } from "./components/ErrorView";
import { useHistory, useParams } from "react-router";

function EditUserForm({ user, onSubmit }) {
  const [firstName, setFirstName] = useState(user.firstName);
  const [lastName, setLastName] = useState(user.lastName);
  const [email, setEmail] = useState(user.email);
  const [password, setPassword] = useState(user.password);
  const history = useHistory();

  async function submit(e) {
    onSubmit(e, { firstName, lastName, email, password });
    history.push("/users");
  }

  return (
    <form onSubmit={submit}>
      <h1>Edit existing user ({firstName})</h1>
      <InputField
        label={"First Name"}
        value={firstName}
        onChangeValue={setFirstName}
      />
      <InputField
        label={"Last Name"}
        value={lastName}
        onChangeValue={setLastName}
      />
      <InputField label={"email"} value={email} onChangeValue={setEmail} />
      <InputField
        type={"password"}
        label={"Password"}
        value={password}
        onChangeValue={setPassword}
      />
      <button>Submit</button>
    </form>
  );
}

export function EditUserPage({ userApi }) {
  const { id } = useParams();

  const { loading, error, data: user, reload } = useLoading(
    async () => await userApi.getUser(id),
    [id]
  );

  async function handleSubmit(e, { firstName, lastName, email, password }) {
    e.preventDefault();
    await userApi.updateUser(id, { firstName, lastName, email, password });
  }

  if (error) return <ErrorView error={error} reload={reload} />;
  if (loading || !user) return <LoadingView />;

  return <EditUserForm user={user} onSubmit={handleSubmit} />;
}
