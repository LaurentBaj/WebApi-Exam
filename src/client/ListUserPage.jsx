import React from "react";
import { LoadingView } from "./components/LoadingView";
import { Link } from "react-router-dom";
import { useLoading } from "./useLoading";
import { ErrorView } from "./components/ErrorView";

export function ListUserPage({ userApi, loadProfile, msg, h5 }) {
  const { data: users, error, loading, reload } = useLoading(
    async () => await userApi.listUsers()
  );

  const { data } = useLoading(async () => await loadProfile());

  if (error) return <ErrorView error={error} reload={reload} />;
  if (loading || !users) return <LoadingView />;

  /* If login add link */
  function output(data, firstName, id, lastName, email, password) {
    if (data)
      return (
        <Link to={`/users/${id}/edit`}>
          User {id}: {firstName} {lastName}
        </Link>
      );
    else return `User ${id}: ${firstName} ${lastName}`;
  }

  return (
    <div>
      <h5>{h5}</h5>
      {users.map(({ id, firstName, lastName }) => (
        <li key={id}>{output(data, firstName, id, lastName)}</li>
      ))}
      <p>{msg}</p>
    </div>
  );
}
