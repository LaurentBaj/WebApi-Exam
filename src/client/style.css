body {
  margin: 0;
  padding: 0;
  font-family: "Cambria Math", serif;
}

#root {
  display: grid;
  grid-template-columns: auto;
  grid-template-rows: auto;
}

header {
  background-color: #004444;
  text-align: center;
  display: grid;
  height: 50px;
  grid-template-columns: 1fr 1fr 1fr 1fr;
}

main {
  text-align: center;
  font-size: 3em;
}

.link {
  margin-top: 2%;
  text-decoration: none;
  color: aliceblue;
}

.output {
  font-size: 1rem;
}

img {
  width: 10%;
  height: 10%;
}

p {
  font-size: 2rem;
}

li {
  font-size: 1.5rem;
}

input {
  height: 50px;
  font-size: 1.5rem;
}

h1 {
  color: #006666;
}

button {
  font-size: 2rem;
}
