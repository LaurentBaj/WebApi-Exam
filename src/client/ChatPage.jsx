import React, { useEffect, useState } from "react";
import { ChatView } from "./ChatView";

export function ChatPage({ member }) {
  const [chatLog, setChatLog] = useState([]);
  const [ws, setWs] = useState();

  useEffect(() => {
    const ws = new WebSocket("ws://localhost:3000");
    ws.onopen = (event) => {
      console.log("opened", event);
    };
    ws.onmessage = (event) => {
      console.log("from server", event);
      setChatLog((chatLog) => [...chatLog, event.data]);
    };
    ws.onclose = (event) => {
      console.log("close", event);
    };
    setWs(ws);
  }, []);

  return (
    <ChatView
      chatLog={chatLog}
      onSendMessage={(message) => ws.send(message)}
      member={member}
    />
  );
}
