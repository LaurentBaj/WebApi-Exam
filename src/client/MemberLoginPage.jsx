import React, { useState } from "react";
import { InputField } from "./components/InputField";
import { ChatPage } from "./ChatPage";

function MemberProfile({ member }) {
  return (
    <div>
      <h3>Logged in as ({member.firstName})</h3>
      <ChatPage member={member} />
    </div>
  );
}

export function MemberLoginPage({ userApi }) {
  const [firstName, setFirstName] = useState();
  const [password, setPassword] = useState();
  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const [member, setMember] = useState("Guest");

  async function submit(e) {
    e.preventDefault();
    const users = await userApi.listUsers();
    const user = users.find((e) => e.firstName === firstName);
    if (firstName === user.firstName && password === user.password) {
      setIsLoggedIn(true);
      setMember(user);
    }
  }

  return (
    <>
      {isLoggedIn && <MemberProfile member={member} />}
      {!isLoggedIn && (
        <form onSubmit={submit}>
          <h1>Login as user</h1>
          <InputField
            label={"Surname"}
            onChangeValue={setFirstName}
            value={firstName}
          />
          <InputField
            label={"Password"}
            onChangeValue={setPassword}
            value={password}
            type={"password"}
          />
          <button>Submit</button>
        </form>
      )}
    </>
  );
}
