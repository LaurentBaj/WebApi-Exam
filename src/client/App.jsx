import React, { useEffect, useState } from "react";
import { BrowserRouter, Link } from "react-router-dom";
import { Route, Switch } from "react-router";

import { ProfilePage } from "./ProfilePage";
import { fetchJSON, fetchJson, postJSON } from "./lib/http";
import { LoginPage } from "./LoginPage";
import { LoginCallbackPage } from "./LoginCallbackPage";
import { ListUserPage } from "./ListUserPage";
import { EditUserPage } from "./EditUserPage";
import { CreateUserPage } from "./CreateUserPage";
import { MemberLoginPage } from "./MemberLoginPage";

function useLocalStorage(key) {
  const [value, setValue] = useState(() =>
    JSON.parse(localStorage.getItem(key))
  );

  useEffect(() => {
    if (value) {
      localStorage.setItem(key, JSON.stringify(value));
    } else {
      localStorage.removeItem(key);
    }
  }, [value]);
  return [value, setValue];
}

export function App() {
  const userApi = {
    listUsers: async () => await fetchJSON("/api/users"),
    getUser: async (id) => await fetchJSON(`/api/users/${id}`),
    createUser: async ({ firstName, lastName, email, password }) => {
      return postJSON("/api/users", {
        json: { firstName, lastName, email, password },
        method: "POST",
      });
    },
    updateUser: async (id, { firstName, lastName, email, password }) =>
      postJSON(`/api/users/${id}`, {
        json: { firstName, lastName, email, password },
        method: "PUT",
      }),
  };

  const [access_token, setAccess_token] = useLocalStorage("access_token");

  const googleIdentityProvider = {
    discoveryURL:
      "https://accounts.google.com/.well-known/openid-configuration",
    client_id:
      "243023781863-mtmgurofsoc7hv9av9q53onkg5lo4dh6.apps.googleusercontent.com",
    scope: "openid email profile",
  };

  async function loadProfile() {
    return fetchJson("/api/profile", {
      headers: {
        ...(access_token ? { Authorization: `Bearer ${access_token}` } : {}),
      },
    });
  }

  return (
    <div id={"container"}>
      <BrowserRouter>
        {/*Front page - Components*/}
        <header>
          <Link className={"link"} to={"/profile"}>
            Profile
          </Link>
          <Link className={"link"} to={"/login"}>
            Admin
          </Link>
          <Link className={"link"} to={"/"}>
            Front Page
          </Link>
          <Link className={"link"} to={"/member"}>
            Login
          </Link>
        </header>
        <main>
          <Switch>
            <Route exact path={"/"} id={"container"}>
              <h1>CoronaBook</h1>
              <ListUserPage userApi={userApi} h5={"Current Members"} />
            </Route>

            {/*Login - Components*/}
            <Route path={"/profile"}>
              <ProfilePage loadProfile={loadProfile} />
            </Route>
            <Route exact path={"/login"}>
              <LoginPage identityProvider={googleIdentityProvider} />
            </Route>
            <Route path={"/login/callback"}>
              <LoginCallbackPage
                identityProvider={googleIdentityProvider}
                onAccessToken={(access_token) => setAccess_token(access_token)}
              />
            </Route>

            {/*Cantine - Components*/}
            <Route path={"/create"}>
              <CreateUserPage userApi={userApi} />
            </Route>
            <Route exact path={"/users"}>
              <ListUserPage
                loadProfile={loadProfile}
                userApi={userApi}
                msg={"Click on the users to edit them"}
                h5={"Edit Page"}
              />
            </Route>
            <Route path={"/users/:id/edit"}>
              <EditUserPage userApi={userApi} />
            </Route>
            <Route path="/member">
              <MemberLoginPage userApi={userApi} />
            </Route>
            <Route>
              <h1>Page not found</h1>
            </Route>
          </Switch>
        </main>
      </BrowserRouter>
    </div>
  );
}
