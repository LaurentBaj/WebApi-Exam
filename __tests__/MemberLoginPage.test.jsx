import React from "react"
import {describe, it} from "@jest/globals";
import * as ReactDOM from "react-dom";
import {MemoryRouter} from "react-router-dom";
import {MemberLoginPage} from "../src/client/MemberLoginPage";


const member = {firstName: "Laurent", password:"1234"}

describe("MemberLoginPage", () => {
    it("render page", async () => {
        const container = document.createElement("div");
        ReactDOM.render(
            <MemoryRouter>
                <MemberLoginPage userApi={member}/>
            </MemoryRouter>,
            container
        );
        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h1").textContent).toEqual(
            "Login as user"
        );
        expect(container.querySelector("label").textContent).toEqual(
            "Surname: "
        );
    });
})

