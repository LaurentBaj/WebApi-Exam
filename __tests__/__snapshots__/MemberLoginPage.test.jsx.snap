// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberLoginPage render page 1`] = `"<form><h1>Login as user</h1><div><label>Surname: <input type=\\"text\\" value=\\"\\"></label></div><div><label>Password: <input type=\\"password\\" value=\\"\\"></label></div><button>Submit</button></form>"`;
